### ISSUE TITLE
getting the current library release via API

### ISSUE THREAD

**Issue by gaiaclary** [KhronosGroup/OpenCOLLADA#531]
Hi;
The Blender executable can produce the list of external library versions that have been used for a specific build. We want to include OpenCollada to that list, that's why i am looking for an API call to get the version number during runtime. I was looking  for something similar to these hypothetical functions:

   static std::string ColladaBUUtils::get_library_version_string(){...}
   static int[3] ColladaBUUtils::get_library_version_info(){...};

But i could not find anything like that. If there is such a function could you please point me to it? or if it does not exist would it be easy to implement? I have seen something is setup in the main CMakeLists.txt but that looks like the version string is always set to "0.1.0" which does not make too much sense.

Any hint would be very appreciated.
thanks

cheers,
Gaia

**Comment by gaiaclary** [comment]
Sergey pointed me to the file:

    https://github.com/KhronosGroup/OpenCOLLADA/blob/master/COLLADABaseUtils/include/COLLADABUVersionInfo.h

However this file contains wrong information and it seems to not be used anyways. At least i can not find it in the includes folder of the installed OPenCollada library folders.

**Comment by RemiArnaud** [comment]
The build process patches the string, but does not commit the changes to master, which stays a developer branch. 
The modified .h is available as artifact in the distributed binaries currently distributed at opencollada.fl4re.com. 
However the master is tagged with the release version if the build and distribution succeeds. The idea is to be able to get the source code for a released version. 

One of the reason we have done this is to avoid untrusted binaries to circulate. We make sure we maintain our build servers with stable environment, and did not want to see a variety of libraries with a non developer build number. 

So, one immediate possibility is for you to get the binaries. Another is to write a small script to get the tag from github and patch the .h in your build process. If you don so, you may want to add additional information to indicate this binary is from the blender build system.

**Comment by LazyDodo** [comment]
Allright couple of issues with that. 

1) The code has a version, every other library you can query that version and get something useful back, regardless of who build it. 

2) We only use the official .zip releases, I have unpacked 1.6.50 and 1.6.51 into different folders, and there's no way to tell which is which version, that's not right... that's not right *at all* . 

3) Blender can't use your libraries, the site linked has only win64 libraries, which appear to be msvc2015/win64/dynamic c runtime libraries. Blender requires the following libs

msvc2013/x64/debug/static crt
msvc2013/x64/release/static crt
msvc2013/x86/debug/static crt
msvc2013/x86/release/static crt
msvc2015/x64/debug/static crt
msvc2015/x64/release/static crt
msvc2015/x86/debug/static crt
msvc2015/x86/release/static crt

however as a rule, we build everything from source, so we wouldn't use your binaries even if they were available. 

4) for an open source project to require your users to use binaries stored on a dropbox is somewhat odd to start with, providing no way to validate those binaries have not been tampered with seems just plain irresponsible. 

5) A nerfarious actor could easily put the version number in, and there will be no way for you to tell if it is a trusted binary or not. 

6) All linux distributions will build from source for their packages, and will therefore also lack the version number.

In conclusion, this is causing a lot of issues and it doesn't seem to be solving any problems, please add the proper version number to the releases in the releases tab and make everyone's life easier.

**Comment by RemiArnaud** [comment]
Yes, make sense to have the correct version number in the source package available in the release tab. I am not sure how to do that

We have a task opened for ages to use github release to store the various binaries instead of using Dropbox, maybe the patched source package can be stored there as well. 

Would this work? Do you have another idea?

**Comment by gaiaclary** [comment]
Maybe there is a way that can make both sides happy:

What about always providing a release number so that any build also from third parties contains it and makes a distribution identifiable as containing a specific release build. But in addition provide something that marks the builds made by OPenCollada "official build" so that the user can get the version information and the origin of the build as separate information ?

This would make us happy because we have the version number and it would make OPenCollada happy because you can distinguish between "official" builds and "third party builds".

**Comment by LazyDodo** [comment]
It's pretty common when a library does a release, they branch and manually update the version number in the header file.

**Comment by gaiaclary** [comment]
I have been doing this sort of things for our own work as follows:

Whenever we publish a new release we call a shell script, e.g like fro example:

  ./makerelease 1.6.51

- The script updates (patches) the version file from the given input parameter
- the version file is committed
- the versio tag is generated from the given input parameter
- the commit and the tag are published to the repository

This works pretty easy for our purposes. However there may be more clever solutions for this. But that is an easy one :)

**Comment by RemiArnaud** [comment]
We could branch and push the updated .h in a release branch rather than just tagging the master branch.

Or we could push the sources in GitHub release with the patched include file. 

But I would like to keep the master branch unchanged. 

@krunal-patel-fl4re @mbarnes-sb  can you comment please?

**Comment by krunal-patel-fl4re** [comment]
Probably best to do all of tag, release branch, and GH release.  The release branch name would be of the form release/X.Y and the updated .h file will store X.Y.Z, and a patch release would just update that release branch with a Z+1 value.

**Comment by RemiArnaud** [comment]
@krunal-patel-fl4re updated the release scripts and the first step is done. we now release the binaries directly on github. See https://github.com/KhronosGroup/OpenCOLLADA/releases

the sources are there too, but still without the version number patched.  This is next on the todo list.

**Comment by RemiArnaud** [comment]
I'm please to announce that we now package source code with the release number burned into the include file. Available now in https://github.com/KhronosGroup/OpenCOLLADA/releases/tag/v1.6.59

**Comment by gaiaclary** [comment]
Hi;
Where exactly do i find the release number now? 
I looked in COLLADABUVersionInfo.h but there i still see 

    const String CURRENT_REVISION = "1.6.developer_version";

I looked into:

    https://github.com/KhronosGroup/OpenCOLLADA/blob/v1.6.59/COLLADABaseUtils/include/COLLADABUVersionInfo.h

**Comment by LazyDodo** [comment]
I grabbed the source release from that tag and can confirm it indeed still says 1.6.developer_version.

**Comment by RemiArnaud** [comment]
Did you get the v1.6.59-Source from https://github.com/KhronosGroup/OpenCOLLADA/releases ?

Just above the 'downloads' section. 

Direct link is https://s3-us-west-2.amazonaws.com/opencollada-promotion-artifacts/OpenCOLLADA_v1.6.59.zip

You can replace the release number with the release you need (but 1.6.59 is the first available)

**Comment by LazyDodo** [comment]
Yeah that zip file has the correct version, i (foolishly?) clicked on the source download right under the 'downloads' section which does not have the version fixed. 

Now we're in a situation where the source files provided on amazon are different than the source files linked in the release and in the 1.6.59 tag.  While not ideal it's still better than what we had, if you would  extend the release script a tiny bit more, so it'll also report the hashes for the archives it generates for the amazon bucket so we could detect post release tampering that would be deeply appreciated .

**Comment by gaiaclary** [comment]
What speaks against simply updating and check-in the release version number right into the sources? Then we could just checkout the tagged version directly from the repository and always be sure whatever we checked out it comes with the appropriate release number.

**Comment by RemiArnaud** [comment]
@LazyDodo - can you precisely describe what you want implemented? How to calculate the tags? Should the tag be in each release slot as plain text comment. Can you open a separate issue for this please - this one is closed. 

@gaiaclary - the bot does not add/comit/push code. The build # is burned in by the build process. The tree is tagged with the version and only the tag is pushed. The tag is the authoritative information. 

If you want to get everything you need using git only - you can clone, pull the tag you want, and patch the .h with your desired tag info. This way you can add whatever you want in the .h and you can check the build was made on your controlled environnement.  Alternatively, you can download a pre-build version that already has the version number burned in.
